packages/flutter/lib/src/widgets/selection.dart

```diff
@@ -0,0 +1,977 @@
+  }
+
+  @override
+  State<SelectionContainer> createState() => _SelectionContainerState();
+}
+
+class _SelectionContainerState extends State<SelectionContainer> implements SelectionRegistrar {
+  final Set<Selectable> _registeredSelectables = <Selectable>{};
+
+  @override
+  void dispose() {
+    widget.delegate?.dispatchSelectionEvent(const ClearSelectionEvent());
+    _registeredSelectables.clear();
+    super.dispose();
+  }
+
+  bool _notifyScheduled = false;
+  final ValueNotifier<bool> _hasSelectableContent = ValueNotifier<bool>(false);
+
+  void _scheduleNotifyListenerIfNeeded() {
+    if (_notifyScheduled)
+      return;
+    _notifyScheduled = true;
+    // We can't notify right away because the Selectables are added during the
+    // build phase and they can't handle selection event until they are laid out
+    // because they don't have size yet.
+    SchedulerBinding.instance!.addPostFrameCallback((Duration timeStamp) {
+      _notifyScheduled = false;
+      widget.delegate!.onSelectablesChange(_registeredSelectables);
+    });
+  }
+
+  @override
+  void add(Selectable selectable) {
+    if (_registeredSelectables.contains(selectable))
+      return;
+    _registeredSelectables.add(selectable);
+    _hasSelectableContent.value = _registeredSelectables.isNotEmpty;
+    _scheduleNotifyListenerIfNeeded();
+  }
+
+  @override
+  void remove(Selectable selectable) {
+    _registeredSelectables.remove(selectable);
+    selectable.dispatchSelectionEvent(const ClearSelectionEvent());
+    _hasSelectableContent.value = _registeredSelectables.isNotEmpty;
+    _scheduleNotifyListenerIfNeeded();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return _SelectableContainerWidget(
+      delegate: widget.delegate,
+      registrar: widget.registrar,
+      hasSelectableContent: _hasSelectableContent,
+      child: _SelectionRegistrarScope(
+        registrar: this,
+        enabled: widget.registrar != null,
+        child: widget.child,
+      ),
+    );
+  }
+}
+
+class _SelectableContainerWidget extends SingleChildRenderObjectWidget {
+  const _SelectableContainerWidget({
+    Key? key,
+    required this.delegate,
+    required this.hasSelectableContent,
+    required this.registrar,
+    required Widget child
+  }) : super(key: key, child: child);
+
+  final SelectionContainerDelegate? delegate;
+  final ValueNotifier<bool> hasSelectableContent;
+  final SelectionRegistrar? registrar;
+
+  @override
+  _RenderSelectableContainerWidget createRenderObject(BuildContext context) {
+    return _RenderSelectableContainerWidget(
+      delegate: delegate,
+      containerHasSelectableContent: hasSelectableContent,
+      selectionRegistrar: registrar,
+    );
+  }
+
+  @override
+  _RenderSelectableContainerWidget updateRenderObject(BuildContext context, _RenderSelectableContainerWidget renderObject) {
+    assert(renderObject.delegate == delegate && renderObject.containerHasSelectableContent == hasSelectableContent);
```

Comment:
What guarantees that these don't change?

Reply:
good catch